controllers/drplacementcontrol.go

```diff
@@ -154,12 +154,6 @@ func (d *DRPCInstance) RunInitialDeployment() (bool, error) {
 			return !done, err
 		}
 
-		d.setDRPCCondition(&d.instance.Status.Conditions, rmn.ConditionAvailable, d.instance.Generation,
```

Comment:
The first reconcile was setting the Available to `true` with this change, the second reconcile will update the status at line 173 (new code). Shouldn't we update the status here as well? (keeping the message constant)

The actual status update would happen in `startProcessing` calling `shouldUpdateStatus`

Reply:
So your concern is "_in case the next reconciliation takes longer than desired_", right?
I tried to reduce code duplication when possible, but, If that's the concern, then, setting up VolSync would also be an issue. We don't set up VolSync until the subsequent reconciliation, which is after the PlacementRule had been updated.

Reply:
For Volsync the Available is not true yet in that case, correct? Or, the condition `available` signifies the ManagedCluster is available to place the workload.

My observation was only that in this reconcile we can update the status, so why wait for the next. Time for the next reconcile is almost immediate, as we return `!done`.

Diversion: We possibly need another condition for workload status, than Available. To let the user know workload health, but that is a different topic.

Reply:
This case seems to me a rare condition.  But, there is a possibility that returning `!done`, might end up stalling due to the exponential backoff.  In that case, VolSync will not be set up until the reconciliation runs again. Again, given that this is an initial deployment I don't expect that to happen, but it is a possibility. 

Given that VolSync hasn't been set up yet, a better solution would be to `requeueAfter` one second or so. That guarantees us that we are in control over VolSync destination setup timing before handing the control over to the exponential backup algorithm.

Reply:
I am struggling to understand why we reconcile based on time, we should rely on dependent objects or watches triggering the reconcile. The exponential backoff is meant to retry but less aggressively as time goes by, which we should honor than circumvent.

Reply:
From `gchat`
---------------------
Shyam Ranganathan, 1:42 PM, Edited
@Benamar Mekhissi PR https://github.com/RamenDR/ramen/pull/522 adds a 5s requeue for volsync DataProtected. Why does the reconciler not reconcile VRG due to a change in child resources? (I forgot the reason)

Benamar Mekhissi, 1:47 PM
Right. With VolSync, we declare that Data is protected once we have at least one sync successful.  Today, we don't watch for VolSync resources changes, hence, that's the reason.

Shyam Ranganathan, 1:47 PM
Then that should just exponentially backoff, right? why tight loop

(I thought VRG owned the Volsync resource, as well)

Benamar Mekhissi, 1:50 PM
Well, in that case, it is possible to take up to 17 minutes right?

Shyam Ranganathan, 1:50 PM
We limited it to 5 minutes from the 17 (or so I thought). But that is the point of backoff, no?

Benamar Mekhissi, 1:51 PM
That's the point, but not for this case.

We want to be able to know that the data is protected.  If are looking at the condition, then we don't.

In the end, things will work just fine, but for users who are watching for the condition, it would be a misinformation

Shyam Ranganathan, 58 min
Hang on... the condition would be updated once it is detected as ready, which may happen at 5 minutes post the last reconcile, right?

Benamar Mekhissi, 56 min
If the reconciliation runs at max 5 minutes

I think the rate limiter that we added is for failures.

Shyam Ranganathan, 53 min
We put in a max delay of 1 minute (I thought that was 5)

Benamar Mekhissi, 53 min
Well again, that's if you return a failure, I think

Shyam Ranganathan, 53 min
It should be for any requeue...

Benamar Mekhissi, 53 min
That's not what's happening in practice

Shyam Ranganathan, 52 min
Fixing that would be better right?

(or if not fixable etc. i.e more a framework problem)

Benamar Mekhissi, 52 min
It depends.  Choosing 1 minutes, 5 minutes, 10???

Benamar Mekhissi, 51 min, Edited
Fixing it right is to just watch for the resource, but that's not the plan right now.

Shyam Ranganathan, 50 min
This fix will do tighter reconciles, I worry we choose that option too frequently

Benamar Mekhissi, 50 min
That's only if VolSync is in play right

Let me check

Shyam Ranganathan, 50 min
^^^ for this PR true, only VolSync

Benamar Mekhissi, 49 min
Right.

Shyam Ranganathan, 49 min
So which resource is this, I remenber VRG owns the RD/RS, so this is something outside those resources we are waiting for?

Benamar Mekhissi, 49 min
So it is the RD and RS.

Benamar Mekhissi, 46 min
So are you saying that if the status of the RD changes, and the owner is VRG, then it will reconcile?

Shyam Ranganathan, 46 min
That is my expectation (managed by, instead of owner maybe), happens with VR to my knowledge

Benamar Mekhissi, 45 min
Because the owner is the VRG, but we get no reconciliation

Shyam Ranganathan, 42 min
Well once we set VR as Primary|Secondary we do not requeue, we let the VR changes tickle upwards: https://github.com/RamenDR/ramen/blob/e2642396a52fe1e1cbbc448d6d63c735244d5b44/controllers/vrg_volrep.go#L978-L983

It seems like the same thing for RS/RD, but I guess the question is if a reconcile happens, which I think happens for VR

Benamar Mekhissi, 40 min
right --- https://github.com/RamenDR/ramen/blob/e2642396a52fe1e1cbbc448d6d63c735244d5b44/controllers/volsync/vshandler.go#L403

Shyam Ranganathan, 39 min
This is what the API States: https://github.com/kubernetes-sigs/controller-runtime/blob/bf71fc56485f6bf03e95ef6b0233ff36c695d4c9/alias.go#L117-L122

// This is used for garbage collection of the owned object and for
// reconciling the owner object on changes to owned (with a Watch + EnqueueRequestForOwner).

Now that last bit about Watch + Enqueue I am not sure is a difference, checking

Benamar Mekhissi, 38 min
Right. We see no reconciliation once the RD is updated

Shyam Ranganathan, 36 min
This adds further ownership for VR: https://github.com/RamenDR/ramen/blob/e2642396a52fe1e1cbbc448d6d63c735244d5b44/controllers/volumereplicationgroup_controller.go#L99

Shyam Ranganathan, 36 min, Edited
I think the above is what is missing for RD/RS causing no reconciles

As that would then no longer depend on exponential backoff, but respond relatively immediately to changes in RD/RD. Question is if we can set Owns for RD/RS...

Benamar Mekhissi, 34 min
We can try it, but then we should also try it when VolSync CRDs are not installed in the MC

So that's something that we wanted to evade.

Shyam Ranganathan, 33 min
Latter is what I am thinking about as well

Current addition to code if avoidable is better, as that is quite a bit in the long run. If we can add the CRDs or conditionally add the Owns it would be better.

I think the discussion was:
- Detect the CRDs and add Owns appropriately
- Watch for the CRDs (or detect VRGs that require Volsync and detect the CRDs) and restart the operator if found, such that it adds the Owns in this incarnation

Benamar Mekhissi, 29 min
Yeah, the goal always was to solve that problem of watching volsync resources later. Once VolSync has been adapted, then at that point, there is no reason not to install the CRDs.

Benamar Mekhissi, 25 min
Ramen/VolSync can live without that change.  In the end, it seems to do the right thing and I didn't want to dig deeper given that we fixed it in that PR. But what we can do in the mean time, we can change that to 30 seconds instead of 5.  Again, that will stop as soon as our data is protected.

Reply:
Done